HaxeManual/08-compiler-features.tex

```diff
@@ -759,3 +759,148 @@ \section{Loop unrolling}
+
+\begin{lstlisting}
+--macro nullSafety("some.package", Off)
+--macro nullSafety("some.package", Loose)
+--macro nullSafety("some.package", Strict)
+\end{lstlisting}
+
+\paragraph{Detailed Usage}
+
+\begin{itemize}
+	\item Null safety makes sure you will not pass nullable values to the places which are not explicitly declared with \type{Null<T>} (assignments, return statements, array access, etc.).
+	\haxe{assets/NullSafety1.hx}
+	\item Using nullables with unary and binary operators (except \expr{==} and \expr{!=}) is not allowed.
+	\item If a field is declared without \type{Null<>} then it should have an initial value or it should be initialized in the constructor (for instance fields).
+	\item Passing an instance of a parametrized type with nullable type parameters where the same type with non-nullable type parameters is expected is not allowed:
+	\begin{lstlisting}
+var nullables:Array<Null<String>> = ['hello', null, 'world'];
+// Array<Null<String>> cannot be assigned to Array<String>:
+//var a:Array<String> = nullables;
+\end{lstlisting}
+	\item Local variables checked against \expr{null} are considered safe inside of a scope covered with that null-check:
+	\begin{lstlisting}
+var nullable:Null<String> = getSomeStr();
+//var s:String = nullable; // Compilation error
+if (nullable != null) {
+  s = nullable; //OK
+}
+//s = nullable; // Compilation error
+s = (nullable == null ? 'hello' : nullable); // OK
+switch (nullable) {
+  case null:
+  case _: s = nullable; // OK
+}
+\end{lstlisting}
+	\item Control flow is also taken into account:
+	\begin{lstlisting}
+function doStuff(a:Null<String>) {
+  if(a == null) {
+    return;
+  }
+  // From here `a` is safe, because function execution will continue only if `a` is not null:
+  var s:String = a; // OK
+}
+\end{lstlisting}
+\end{itemize}
+
+\paragraph{Limitations}
+
+\begin{itemize}
+	\item Out-of-bounds array reads return \expr{null}, but Haxe types them without \type{Null<>}.
+	\begin{lstlisting}
+var a:Array<String> = ["hello"];
+$type(a[100]); // String
+trace(a[100]); // null
+var s:String = a[100]; // Safety does not complain here, because `a[100]` is not `Null<String>`, but just `String`
+\end{lstlisting}
+	\item Out-of-bounds array writes fill all positions between the last defined index and the newly-written one with \expr{null} values. Null safety cannot protect against this.
+	\begin{lstlisting}
+var a:Array<String> = ["hello"];
+a[2] = "world";
+trace(a); // ["hello", null, "world"]
+var s:String = a[1]; // Cannot check this
+trace(s); //null
+\end{lstlisting}
+	\item Haxe was not designed with null safety in mind, so it's always possible \expr{null} values will come into your code from third-party code or even from the standard library.
+	\item Nullable fields and properties are not considered null-safe even after checking against \expr{null}. You can use helper methods instead (e.g. from \href{https://github.com/RealyUniqueName/Safety/}{this library}):
```

Comment:
I'm not sure it's ok to mention 3rd-party libraries in the manual.
Maybe it's better to provide an example code of such a helper method.